### PR TITLE
Introduce the Device Manufacturing Database, and private key extraction support

### LINKED ISSUE #77: Export device data
cmprovision https://github.com/raspberrypi/cmprovision/ has the ability to export data.

It would be nice to get a CSV of every device seen with a success or failure status.

### LINKED ISSUE #79: Decryption of provisioned device
Hi,

If I use this provision a device, and then the device malfunctions and I want to investigate it by decrypting the data - what's the process to decrypt it?  

Thank you!

### PR DESCRIPTION
**Pull request by tdewey-rpi**
Introduce support for a CSV device manufacturing database, and support for extracting device private keys as part of initial manufacturing to support device RMA and recovery scenarios.

Fixes #77 
Fixes #79

### CODE REVIEW

File: service/rpi-sb-provisioner.sh [hunk 1]
```diff
@@ -497,6 +509,13 @@ if [ -z "${DEMO_MODE_ONLY}" ] && [ -n "${RPI_DEVICE_FETCH_METADATA}" ]; then
     esac
 
     keywriter_log "Board is: ${BOARD_STR}, with revision number ${REVISION}. Has Processor ${PROCESSOR_STR} with Memory ${MEMORY_STR}. Was manufactured by ${MANUFACTURER_STR}"
+
+    if [ -n "${RPI_DEVICE_METADATA_CSV}" ]; then
+        if [ ! -e "${RPI_DEVICE_METADATA_CSV}" ]; then
+            echo "serial,keyhash,mac,jtag_locked,advanced_boot,boot_rom,board_attr,boardname,boardrev,processor,memory,manufacturer" > "${RPI_DEVICE_METADATA_CSV}"
```
Comment (roliver-rpi) [inline service/rpi-sb-provisioner.sh:515]: Do we need to consider multiple concurrent access/locking to/for this file?

rpi-sb-provisioner@.service is an instanced service unit and nothing (that I'm aware of) prevents triage.sh from running multiple instances simultaneously. The paths for our other file writes are typically parametrised with ${TARGET_DEVICE_SERIAL} so this hasn't been an issue previously.

File: README.adoc [hunk 2]
```diff
@@ -171,6 +178,15 @@ The metadata is inserted into the log for the device, and also as a serial-numbe
 
 Set to any value to enable metadata collection.
 
+=== RPI_DEVICE_RETRIEVE_PRIVATE_KEY
+*Optional*
```
Comment (torntrousers) [inline README.adoc:182]: mention this is not possible if CUSTOMER_KEY_PKCS11_NAME  is used?

Comment (tdewey-rpi) [reply README.adoc:182]: > oh. is the device unique key the symmetric key used for the LUKSv2 encryption? Having _PRIVATE_ in the name makes it sound like an asymmetric key.

The device unique key is indeed asymmetric - but only the private component is retained by the device. This component is used as the unlock passphrase for LUKS - but can also be used for other purposes by trusted software, where having the public key stored may allow you to authenticate a given device.

**Comment by roliver-rpi** [comment]
Hi @tdewey-rpi , the branch has been force-pushed but Github doesn't seem to have updated the PR for some reason?

File: service/rpi-sb-provisioner.sh (hunk 1 above)
Comment (tdewey-rpi) [reply service/rpi-sb-provisioner.sh]: I swapped this to using a WAL sqlite3 database, which I believe should be resilient against this problem.

File: config/config_app.helper [hunk 3]
```diff
@@ -6,6 +6,8 @@ RPI_DEVICE_FAMILY|Mandatory|Specify the family of Raspberry Pi device you are pr
...
 RPI_DEVICE_LOCK_JTAG|Optional|Raspberry Pi devices have a mechanism to restrict JTAG access to the device. Note that using this function will prevent Raspberry Pi engineers from being able to assist in debugging your device, should you request assitance. Set to any value to enable the JTAG restrictions.
 RPI_DEVICE_EEPROM_WP_SET|Optional|Raspberry Pi devices that use an EEPROM as part of their boot flow can configure that EEPROM to enable write protection - preventing modification. Set to any value to enable EEPROM write protection.
+RPI_DEVICE_METADATA_CSV|Optional|Store manufacturing data in a comma separated values (CSV) file. This will include the board serial, board revision, the boot ROM version, the MAC address of the ethernet port, any set hash of the customer signing key, the JTAG lock state, the board attributes and the advanced boot flags.
```
Comment (roliver-rpi) [inline config/config_app.helper:9]: This is stale, should be RPI_SB_PROVISONER_MANUFACTURING_DB now?

File: config/validator.py [hunk 4]
```diff
@@ -64,15 +64,18 @@ def validate_RPI_DEVICE_LOCK_JTAG(text) -> tuple[bool, str]:
 def validate_RPI_DEVICE_EEPROM_WP_SET(text) -> tuple[bool, str]:
     return (True, "")
 
+def validate_RPI_DEVICE_METADATA_CSV(text) -> tuple[bool, str]:
```
Comment (roliver-rpi) [inline config/validator.py:67]: Also stale?

File: host-support/rpi-sb-provisioner [hunk 5]
```diff
@@ -6,6 +6,8 @@ RPI_DEVICE_FAMILY=
 RPI_DEVICE_BOOTLOADER_CONFIG_FILE=/var/lib/rpi-sb-provisioner/bootloader.config
 RPI_DEVICE_LOCK_JTAG=
 RPI_DEVICE_EEPROM_WP_SET=
+RPI_DEVICE_METADATA_CSV=
```
Comment (roliver-rpi) [inline host-support/rpi-sb-provisioner:9]: Here too?

File: README.adoc [hunk 6]
```diff
@@ -162,6 +162,13 @@ Raspberry Pi devices that use an EEPROM as part of their boot flow can configure
 
 Set to any value to enable EEPROM write protection.
 
+=== RPI_SB_PROVISONER_MANUFACTURING_DB
+*Optional*
+
+Store manufacturing data in a sqlite3 database. This will include the board serial, board revision, the boot ROM version, the MAC address of the ethernet port, any set hash of the customer signing key, the JTAG lock state, the board attributes and the advanced boot flags.
+
+Set to a writeable file path to enable data storage.
```
Comment (roliver-rpi) [inline README.adoc:170]: Should document that this must not be on a network filesystem as WAL is being used.

File: service/rpi-sb-provisioner.sh [hunk 7]
```diff
@@ -497,6 +509,55 @@ if [ -z "${DEMO_MODE_ONLY}" ] && [ -n "${RPI_DEVICE_FETCH_METADATA}" ]; then
     esac
 
     keywriter_log "Board is: ${BOARD_STR}, with revision number ${REVISION}. Has Processor ${PROCESSOR_STR} with Memory ${MEMORY_STR}. Was manufactured by ${MANUFACTURER_STR}"
+
+    if [ -f "${RPI_SB_PROVISONER_MANUFACTURING_DB}" ]; then
```
Comment (roliver-rpi) [inline service/rpi-sb-provisioner.sh:513]: How is it created in the first place?

Comment (tdewey-rpi) [reply service/rpi-sb-provisioner.sh:513]: sqlite automatically creates the backing file, if it does not exist.

Comment (roliver-rpi) [reply service/rpi-sb-provisioner.sh:513]: sqlite isn't invoked outside this 'if [ -f ]' block though? The backing file does not exist, so sqlite is not invoked? Should this be 'if [ ! -z ]' or similar?

Comment (tdewey-rpi) [reply service/rpi-sb-provisioner.sh:513]: Not convinced that would work - among other things, that might lead to a multi-content database, or an existing database becoming a WAL object.

Probably the safest thing to do is to let this error out, and include guidance that a pre-processing step is required.

**Comment by tdewey-rpi** [comment]
Confirmed in testing that this does the right thing, provided you've read the docs. I will explore sqlite3 as a backing store for other state, too.

**Pull request merged**